src/classes/OPP_AutomatedSoftCreditsService.cls

```diff
@@ -238,56 +242,33 @@ public class OPP_AutomatedSoftCreditsService {
-                if (!cleanedRelationships.containsKey(eachRelationship.npe4__RelatedContact__c)) {
-                    cleanedRelationships.put(eachRelationship.npe4__RelatedContact__c, eachRelationship);
-                    uniqueRelationships.add(eachRelationship);
-                }
-            }
-
-            primaryContactToUniqueRelationships.put(primaryContactId, uniqueRelationships);
-        }
-
-        return primaryContactToUniqueRelationships;
-    }
-
-
-    /*********************************************************************************************************
-    * @description Removes duplicate Affiliation records based on the Contact Id.
-    * @param accountToAffiliations Map of Account to Affiliation records.
-    * @return Map<Id, List<npe5__Affiliation__c>> Map of unique Affiliation records based on Contact Id.
-    **********************************************************************************************************/
-    @testVisible
-    private static Map<Id, List<npe5__Affiliation__c>> removeDuplicateAffiliations(Map<Id, List<npe5__Affiliation__c>> accountToAffiliations) {
-        Map<Id, List<npe5__Affiliation__c>> accountToUniqueAffiliations = new Map<Id, List<npe5__Affiliation__c>>();
-
-        for (Id accountId: accountToAffiliations.keySet()) {
-            Map<Id, npe5__Affiliation__c> cleanedAffiliations = new Map<Id, npe5__Affiliation__c>();
-            List<npe5__Affiliation__c> uniqueAffiliations = new List<npe5__Affiliation__c>();
-            List<npe5__Affiliation__c> affiliations = accountToAffiliations.get(accountId);
-            for (npe5__Affiliation__c eachAffiliation : affiliations) {
-                if (!cleanedAffiliations.containsKey(eachAffiliation.npe5__Contact__c)) {
-                    cleanedAffiliations.put(eachAffiliation.npe5__Contact__c, eachAffiliation);
-                    uniqueAffiliations.add(eachAffiliation);
+    private static Map<Id, List<sObject>> removeDuplicateAssociatedRecords(Map<Id, List<sObject>> sourceIdToAssociatedRecords, String objectType, String fieldName) {
+        Map<Id, List<sObject>> sourceIdToUniqueAssociatedRecords = new Map<Id, List<sObject>>();
+
+        for (Id sourceId : sourceIdToAssociatedRecords.keySet()) {
+            Map<Id, sObject> cleanedRelatedRecords = new Map<Id, sObject>();
+            List<sObject> uniqueAssociatedRecords = new List<sObject>();
+            List<sObject> relatedRecords = sourceIdToAssociatedRecords.get(sourceId);
+            for (sObject eachAssociatedRecord : relatedRecords) {
+                if (!cleanedRelatedRecords.containsKey((Id)eachAssociatedRecord.get(fieldName))) {
+                    cleanedRelatedRecords.put((Id)eachAssociatedRecord.get(fieldName), eachAssociatedRecord);
+                    if (OBJECT_TYPE_RELATIONSHIP.equals(objectType)) {
+                        uniqueAssociatedRecords.add((npe4__Relationship__c)eachAssociatedRecord);
```

Comment:
Since your list of `uniqueAssociatedRecords` is on `SObject`, there's actually no need to split these in in the `if` and cast as the explicit object type. It would be more concise to just add them to the collection as an Sobject: `uniqueAssociatedRecords.add(eachAssociatedRecord);`

Reply:
I'll look into that.